File: backend/src/main/kotlin/AuditLog.kt
package com.scienta

import com.google.cloud.firestore.FieldValue

suspend fun writeAuditLog(
    adminUid: String,
    adminEmail: String = "",
    action: String,
    targetType: String,
    targetId: String,
    details: Map<String, Any> = emptyMap()
) {
    val firestore = getFirestore()
    val entry = mutableMapOf<String, Any>(
        "timestamp" to FieldValue.serverTimestamp(),
        "adminUid" to adminUid,
        "adminEmail" to adminEmail,
        "action" to action,
        "targetType" to targetType,
        "targetId" to targetId
    )
    if (details.isNotEmpty()) entry["details"] = details
    firestore.collection("auditLog").add(entry).get()
}

File: frontend/src/lib/auditLog.ts
import { collection, addDoc, serverTimestamp } from 'firebase/firestore'
import { db, auth } from '../firebase'

export async function writeAuditLog(
  action: string,
  targetType: string,
  targetId: string,
  details: Record<string, unknown> = {}
) {
  const user = auth.currentUser
  if (!user) return
  try {
    await addDoc(collection(db, 'auditLog'), {
      timestamp: serverTimestamp(),
      adminUid: user.uid,
      adminEmail: user.email ?? '',
      action,
      targetType,
      targetId,
      details,
    })
  } catch (e) {
    console.error('Failed to write audit log:', e)
  }
}
